### PR TITLE
Add other callsign formats and improve current format

### LINKED ISSUE #1833: [Feature Request/Correction] Add other callsign formats and improve current format
While experimenting with some speech responder functions the other day, I noticed that EDDI doesn't provide a way to get other callsign formats. `ShipCallsign()` also does not match the ingame default.

I checked all 3 ingame options and they are as follows:

- Commander Name: last word of manufacturer name + first 3 alphanumeric characters of commander name
- Ship Name: last word of manufacturer name + first 3 alphanumeric characters of ship name
- Ship ID: last word of manufacturer name + first 3 alphanumeric characters of ship identifier

If the ship name, commander name, or ship identifier are shorter than 3 alphanumeric characters, they are zero-padded.

Examples:

Examples are with the ingame option set to Ship ID.

| Manufacturer       | Ship Model      | Ship Name   |Ship ID | Callsign                    | 
|--------------------|-----------------|-------------|--------|-----------------------------|
| Gutamaya           | Imperial Eagle  | Crying Kite | C      | Gutamaya Charlie Zero Zero  |
| Gutamaya           | Imperial Eagle  | Crying Kite | CRKIE  | Gutamaya Charlie Romeo Kilo |
| Faulcon DeLacy     | Sidewinder      | Sidewinder  | AN-19S | DeLacy Alpha November One   |
| Faulcon DeLacy     | Sidewinder      | Sidewinder  | A      | DeLacy Alpha Zero Zero      |

To make the current callsign function match the ingame, it would need to check for multiple words in the manufacturer name and only take the last one. I'm not sure how your IPA setup works, so this may require a little extra work.

Adding the ship name and ship id callsigns is easy enough, just replace the `EDDI.Instance.Cmdr.name` with the appropriate value, as it works the same way.

Lastly the code needs to check the length of the trimmed name/id and zero-pad if it is shorter than 3 characters.

Obviously there needs to be a way to pick between callsign formats. There could be a config option in EDDI, or have the function itself take an argument denoting which format to produce.
Again, I'm not too certain how the codebase works, so this may not be as easy as I think it is, but it appears to me that an integer 0-2 or an Enum type could be used as a second argument. ex. `ShipCallsign(ship.ident, callsign.shipname)` or `ShipCallsign(null, 1)`

My code insight is from reading [ScriptResolver.cs in tag 3.5.3-b3](https://github.com/EDCD/EDDI/blob/Release/3.5.3-b3/SpeechResponder/ScriptResolver.cs#L606)

### PR DESCRIPTION
**Pull request by Tkael**
Resolves #1833.

### CODE REVIEW

File: SpeechResponder/CustomFunctions/ShipCallsign.cs [hunk 1]
```diff
@@ -17,36 +20,87 @@ public class ShipCallsign : ICustomFunction
         public string description => Properties.CustomFunctions_Untranslated.ShipCallsign;
         public NativeFunction function => new NativeFunction((values) =>
         {
-            int? localId = (values.Count == 0 ? (int?)null : (int)values[0].AsNumber);
+            // The game provides three options for callsigns used by in-game ATC:
```
Comment (richardbuckle) [inline SpeechResponder/CustomFunctions/ShipCallsign.cs:24]: LEt's get some test coverage on this, as it's quite complex.

File: SpeechResponder/CustomFunctions/ShipCallsign.cs [hunk 2]
```diff
@@ -17,36 +20,87 @@ public class ShipCallsign : ICustomFunction
...
+                : (int?)null);
+            int? callsignType = (values.Count > 1 && values[1].Type != ValueContent.Void 
+                ? (int)values[1].AsNumber 
+                : (int?)null);
+
             ShipMonitor shipMonitor = (ShipMonitor)EDDI.Instance.ObtainMonitor("Ship monitor");
-            Ship ship = shipMonitor.GetShip(localId);
+            Ship ship = localId is null ? shipMonitor.GetCurrentShip() : shipMonitor.GetShip(localId);
+
+            string result = "";
```
Comment (richardbuckle) [inline SpeechResponder/CustomFunctions/ShipCallsign.cs]: Consider a StringBuilder maybe? But only once the tests are in place ;)
  https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder?view=net-5.0

File: SpeechResponder/CustomFunctions/ShipCallsign.cs [hunk 3]
```diff
@@ -17,36 +20,87 @@ public class ShipCallsign : ICustomFunction
...
-                if (EDDI.Instance.Cmdr != null && EDDI.Instance.Cmdr.name != null)
-                {
-                    // Obtain the first three characters
-                    string chars = new Regex("[^a-zA-Z0-9]").Replace(EDDI.Instance.Cmdr.name, "").ToUpperInvariant().Substring(0, 3);
-                    result = ship.phoneticmanufacturer + " " + Translations.ICAO(chars);
-                }
-                else
-                {
-                    if (string.IsNullOrEmpty(ship.phoneticmanufacturer))
+                case "Core Dynamics":
```
Comment (richardbuckle) [inline SpeechResponder/CustomFunctions/ShipCallsign.cs]: Would it be easier to just add a Dictionary in `ShipDefinitions.cs`, similar to `ManufacturerPhoneticNames`?

File: SpeechResponder/CustomFunctions/ShipCallsign.cs [hunk 4]
```diff
@@ -17,36 +20,87 @@ public class ShipCallsign : ICustomFunction
...
                     }
-                }
+                    break;
             }
-            else
+
+            string Get3LeadingCharacters(string input)
             {
-                result = "unidentified ship";
+                // Obtain the first three characters of the input string, zero padded and converted to ICAO (e.g. "A" becomes "Alpha Zero Zero")
```
Comment (richardbuckle) [inline SpeechResponder/CustomFunctions/ShipCallsign.cs]: Test coverage would be good here too. I can think of plenty of edge cases like "A-1-B" or "--A--".

**Review by richardbuckle — CHANGES_REQUESTED** [review]
Too good but I'd be much happier with test coverage.

File: SpeechResponder/CustomFunctions/ShipCallsign.cs (hunk 1 above)
Comment (Tkael) [reply SpeechResponder/CustomFunctions/ShipCallsign.cs:24]: Done.

File: SpeechResponder/CustomFunctions/ShipCallsign.cs (hunk 2 above)
Comment (Tkael) [reply SpeechResponder/CustomFunctions/ShipCallsign.cs]: Done.

File: SpeechResponder/CustomFunctions/ShipCallsign.cs (hunk 4 above)
Comment (Tkael) [reply SpeechResponder/CustomFunctions/ShipCallsign.cs]: Done.

File: SpeechResponder/CustomFunctions/ShipCallsign.cs (hunk 3 above)
Comment (Tkael) [reply SpeechResponder/CustomFunctions/ShipCallsign.cs]: Perhaps, but this is a rather specialized application isn't it?
